### PR TITLE
Module3

### PR DESCRIPTION
**Pull request by maruvuina**

### CODE REVIEW

File: module2/build.gradle [hunk 1]
```diff
@@ -13,16 +13,20 @@ subprojects {
...
     }
 
     dependencies {
         implementation group: 'org.springframework', name: 'spring-core', version: spring_version
         implementation (group: 'org.springframework', name: 'spring-context', version: spring_version) {
             exclude group: 'commons-logging', module: 'commons-logging'
         }
+        implementation group: 'org.springframework.boot', name: 'spring-boot-starter', version: spring_boot_version
+        implementation group: 'org.springframework.boot', name: 'spring-boot-starter-hateoas', version: spring_boot_version
+        implementation group: 'org.springframework', name: 'spring-orm', version: spring_version
```
Comment (olegborikov) [inline module2/build.gradle:27]: why you need that dependency?

Comment (olegborikov) [reply module2/build.gradle:27]: Use spring-boot-starter-data-jpa

File: module2/dao/build.gradle [hunk 2]
```diff
@@ -1,13 +1,17 @@
 ext {
+    hibernate_version = '5.6.2.Final'
     postgresql_version = '42.3.1'
     annotations_version = '12.0'
 }
 
 dependencies {
-    implementation group: 'org.springframework', name: 'spring-jdbc', version: spring_version
+    implementation group: 'org.hibernate', name: 'hibernate-core', version: hibernate_version
```
Comment (olegborikov) [inline module2/dao/build.gradle:8]: Use spring-boot-starter-data-jpa

File: module2/dao/src/main/java/com/epam/esm/dao/CreateDao.java [hunk 3]
```diff
@@ -0,0 +1,17 @@
+package com.epam.esm.dao;
+
+/**
+ *This is an interface for create operations of entity.
+ *
+ * @param <T> a concrete type of entity
+ */
+public interface CreateDao<T> {
```
Comment (olegborikov) [inline module2/dao/src/main/java/com/epam/esm/dao/CreateDao.java:8]: As for me such decomposition is redundant. You can create CRDDao

File: module2/dao/src/main/java/com/epam/esm/dao/entity/GiftCertificate.java [hunk 4]
```diff
@@ -1,29 +1,76 @@
...
 @AllArgsConstructor
 @NoArgsConstructor
+@EntityListeners({GiftCertificateAuditListener.class})
 public class GiftCertificate {
 
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @Column(name = "id")
     private Long id;
+    @Column(name = "name", nullable = false, unique = true)
```
Comment (olegborikov) [inline module2/dao/src/main/java/com/epam/esm/dao/entity/GiftCertificate.java:43]: You can remove @Column here. Hibernate will parse name automatically

File: module2/dao/src/main/java/com/epam/esm/dao/entity/GiftCertificate.java [hunk 5]
```diff
@@ -1,29 +1,76 @@
...
     private Instant lastUpdateDate;
+    @ManyToMany(cascade = CascadeType.MERGE,
+            fetch = FetchType.EAGER)
+    @JoinTable(name = "gift_certificate_tag",
+            joinColumns = @JoinColumn(name = "gift_certificate_id"),
+            inverseJoinColumns = @JoinColumn(name = "tag_id"))
     private List<Tag> tags;
+
+    @Override
+    public String toString() {
```
Comment (olegborikov) [inline module2/dao/src/main/java/com/epam/esm/dao/entity/GiftCertificate.java:63]: You can use Lombok ToString

File: module2/web/src/main/resources/application.yml [hunk 6]
```diff
@@ -0,0 +1,14 @@
+spring:
+  datasource:
+    driver-class-name: org.postgresql.Driver
+    url: jdbc:postgresql://localhost:5432/gift_certificates_system
+    username: postgres
+    password: root
```
Comment (olegborikov) [inline module2/web/src/main/resources/application.yml:6]: It is better not to store username and password like that. It is sensitive information. You can use ${USERNAME} here and pass values via env vars or command vars

File: module2/web/build.gradle [hunk 7]
```diff
@@ -2,11 +2,14 @@ plugins {
...
+
 dependencies {
     implementation project (':dao')
     implementation project (':service')
-    implementation group: 'org.springframework', name: 'spring-webmvc', version: spring_version
-    implementation group: 'org.springframework', name: 'spring-tx', version: spring_version
-    implementation group: 'com.fasterxml.jackson.core', name: 'jackson-databind', version: '2.13.0'
+    implementation group: 'org.springframework.boot', name: 'spring-boot-starter-web', version: spring_boot_version
+    implementation group: 'org.springframework.boot', name: 'spring-boot-starter-validation', version: spring_boot_version
+    implementation group: 'org.atteo', name: 'evo-inflector', version: '1.3'
```
Comment (olegborikov) [inline module2/web/build.gradle:14]: Why you need that dependency?

File: module2/service/src/main/java/com/epam/esm/service/validator/impl/OrderValidatorImpl.java [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+import com.epam.esm.service.validator.OrderValidator;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Component;
+
+import static com.epam.esm.service.exception.ErrorCode.ERROR_301400;
+import static com.epam.esm.service.exception.ErrorCode.ERROR_302400;
+
+@Slf4j
+@Component
+public class OrderValidatorImpl implements OrderValidator {
```
Comment (olegborikov) [inline module2/service/src/main/java/com/epam/esm/service/validator/impl/OrderValidatorImpl.java:14]: It is a little bit strange that in controller you use Hibernate validation, but here you use custom validation

**Comment by olegborikov** [comment]
root project name is module2 https://github.com/maruvuina/spring-rest-project/blob/module3/module2/settings.gradle#L1

**Review by olegborikov — APPROVED** [review]

**Review by GlebBalashevich — APPROVED** [review]

File: module2/web/src/main/resources/application.yml (hunk 6 above)
Comment (maruvuina) [reply module2/web/src/main/resources/application.yml]: fixed

File: module2/web/build.gradle (hunk 7 above)
Comment (maruvuina) [reply module2/web/build.gradle:14]: It's used to represent hateoas links name in valid form.

File: module2/dao/src/main/java/com/epam/esm/dao/entity/GiftCertificate.java (hunk 5 above)
Comment (maruvuina) [reply module2/dao/src/main/java/com/epam/esm/dao/entity/GiftCertificate.java:62]: fixed

File: module2/dao/src/main/java/com/epam/esm/dao/entity/GiftCertificate.java (hunk 4 above)
Comment (maruvuina) [reply module2/dao/src/main/java/com/epam/esm/dao/entity/GiftCertificate.java]: fixed

File: module2/build.gradle (hunk 1 above)
Comment (maruvuina) [reply module2/build.gradle:23]: fixed

File: module2/dao/build.gradle (hunk 2 above)
Comment (maruvuina) [reply module2/dao/build.gradle:8]: fixed

**Pull request merged**